src/main/scala/is/hail/io/index/IndexReader.scala

```diff
@@ -0,0 +1,157 @@
+package is.hail.io.index
+
+import java.util
+import java.util.Map.Entry
+
+import is.hail.annotations.{Annotation, ExtendedOrdering, RegionValue, SafeRow}
+import is.hail.expr.Parser
+import is.hail.io._
+import is.hail.rvd.RVDContext
+import is.hail.utils._
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.FSDataInputStream
+import org.apache.spark.sql.Row
+import org.json4s.Formats
+import org.json4s.jackson.JsonMethods
+
+class IndexReader(conf: Configuration, path: String, cacheCapacity: Int = 256) {
+  private val is = conf.unsafeReader(path + "/index").asInstanceOf[FSDataInputStream]
+  private val codecSpec = CodecSpec.default
+
+  private val metadata = readMetadata(path + "/metadata.json.gz")
+  val branchingFactor = metadata.branchingFactor
+  val nKeys = metadata.nKeys
+  val attributes = metadata.attributes
+
+  val keyType = Parser.parseType(metadata.keyType)
+  val leafType = LeafNodeBuilder.typ(keyType)
+  val internalType = InternalNodeBuilder.typ(keyType)
+
+  private val leafDecoder = codecSpec.buildDecoder(leafType, leafType)(is)
+  private val internalDecoder = codecSpec.buildDecoder(internalType, internalType)(is)
+
+  private val ctx = RVDContext.default
+  private val region = ctx.region
+  private val rv = RegionValue(region)
+
+  @transient private[this] lazy val leafCache = new util.LinkedHashMap[Long, LeafNode](cacheCapacity, 0.75f, true) {
+    override def removeEldestEntry(eldest: Entry[Long, LeafNode]): Boolean = size() > cacheCapacity
+  }
+
+  private def readMetadata(metadataFile: String): IndexMetadata = {
+    val jv = conf.readFile(metadataFile) { in => JsonMethods.parse(in) }
+    implicit val formats: Formats = defaultJSONFormats
+    jv.extract[IndexMetadata]
+  }
+
+  private def readInternalNode(offset: Long): InternalNode = {
+    is.seek(offset)
+    assert(internalDecoder.readByte() == 1)
+    rv.setOffset(internalDecoder.readRegionValue(region))
+    InternalNode(SafeRow(internalType, rv))
+  }
+
+  private def readLeafNode(offset: Long): LeafNode = {
+    is.seek(offset)
+    assert(leafDecoder.readByte() == 0)
+    rv.setOffset(leafDecoder.readRegionValue(region))
+    LeafNode(SafeRow(leafType, rv))
+  }
+
+  private def queryByKey(key: Annotation, depth: Int, nodeOffset: Long, ordering: ExtendedOrdering): Option[Long] = {
```

Comment:
I have a second branch with queryByKey (removed it in this branch) because the code was more complicated and I felt it was better to be looked at separately.